ml_peg/analysis/physicality/diatomics/analyse_diatomics.py

```diff
@@ -0,0 +1,403 @@
+    """
+    Load diatomic curve data for a model.
+
+    Parameters
+    ----------
+    model_name
+        Model identifier.
+
+    Returns
+    -------
+    pd.DataFrame
+        Dataframe containing pair, distance, energy, and projected force columns.
+    """
+    csv_path = CALC_PATH / model_name / "diatomics.csv"
+    if not csv_path.exists():
+        return pd.DataFrame()
+    return pd.read_csv(csv_path)
+
+
+def prepare_pair_series(
+    pair_dataframe: pd.DataFrame,
+) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
+    """
+    Sort and align energy/force series for a diatomic pair.
+
+    Parameters
+    ----------
+    pair_dataframe
+        Pair-specific dataframe.
+
+    Returns
+    -------
+    tuple[np.ndarray, np.ndarray, np.ndarray]
+        Distances, shifted energies, and projected forces sorted by decreasing
+        distance.
+    """
+    df_sorted = pair_dataframe.sort_values("distance").drop_duplicates("distance")
+    series = df_sorted[["distance", "energy", "force_parallel"]].to_numpy()
+    if len(series) < 3:
+        return np.array([]), np.array([]), np.array([])
+
+    distances = series[:, 0]
+    energies = series[:, 1]
+    forces = series[:, 2]
+
+    descending = np.argsort(distances)[::-1]
+    distances = distances[descending]
+    energies = energies[descending]
+    forces = forces[descending]
+
+    shifted_energies = energies - energies[0]
+    return distances, shifted_energies, forces
+
+
+def count_sign_changes(array: np.ndarray, tol: float) -> int:
+    """
+    Count sign changes in a sequence while ignoring small magnitudes.
+
+    Parameters
+    ----------
+    array
+        Input values.
+    tol
+        Absolute tolerance below which values are treated as zero.
+
+    Returns
+    -------
+    int
+        Number of sign changes exceeding the specified tolerance.
+    """
+    if array.size < 3:
+        return 0
+    clipped = array.copy()
+    clipped = clipped[np.abs(clipped) > tol]
+    signs = np.sign(clipped)
+    sign_flips = signs[:-1] != signs[1:]
+    return int(np.sum(sign_flips))
+
+
+def compute_pair_metrics(
+    df_pair: pd.DataFrame,
+) -> tuple[dict[str, float], float] | tuple[None, None]:
+    """
+    Compute diagnostics for a single diatomic pair.
+
+    Parameters
+    ----------
+    df_pair
+        Pair-specific dataframe.
+
+    Returns
+    -------
+    tuple[dict[str, float], float] | tuple[None, None]
+        Dictionary of metrics and the homonuclear well depth (if applicable).
+    """
+    distances, shifted_energies, projected_forces = prepare_pair_series(df_pair)
+    if distances.size < 3:
+        return None, None
+
+    energy_gradient = np.gradient(shifted_energies, distances)
+    energy_curvature = np.gradient(energy_gradient, distances)
+
+    minima = 0
+    if shifted_energies.size >= 3:
+        second_diff = np.diff(np.sign(np.diff(shifted_energies)))
+        minima = int(np.sum(second_diff > 0))
```

Comment:
This is sensitive to all noise e.g. `shifted_energies = np.cos(np.linspace(0, 30)) + np.random.rand(50)` will roughly double the number of 'minima' compared to the same energies without noise.

This obviously depends on the noise compared to the average energy difference, so maybe isn;t a problem. Do you think this is ok in practice?